### DISCUSSION TITLE
How to link and include zenoh-cpp in another project CMakeLists.txt?

### DISCUSSION THREAD

**Discussion by lucasw** [eclipse-zenoh/roadmap#103] [Q&A] [answered]
Some of the things I've tried are in https://github.com/lucasw/ros_one2z/issues/2 - I think have zenoh-cpp built and installed correctly and have a libzenohc.so and zenohc.hxx, zenoh.h, and zenoh.hxx, some lib/cmake zenoh files (but they don't provide a zenoh_INCLUDE_DIRECTORY or similar?  I had to specify the include path on the cmake command line for my project) in my custom install destination, but I'm not sure what to put in my [CMakeLists.txt](https://github.com/lucasw/ros_one2z/blob/zenoh/ros1_example_pkg/CMakeLists.txt#L35-L38) to find and link to it.

Should it be something like this?

```
find_package(zenohcxx REQUIRED COMPONENTS zenohcxx::zenohc::lib)
...
include_directories(???)
...
target_link_libraries(my_executable PUBLIC zenohcxx::zenohc::lib ${catkin_LIBRARIES})
```

(It seems like zenoh-c and zenoh-cpp ought to install to `include/zenoh/` rather than `include/`, and files should do `#include <zenoh/zenoh.hxx>` instead of `#include <zenoh.hxx>`)

Currently with that I get

```
Target "zenoh_generate_image" links to:

    zenohcxx::zenohc::lib

  but the target was not found.  Possible reasons include:
```

**Comment by milyin** [comment, marked as answer]
The problem seems to be that zenoh-c was not included before. The zenoh-cpp is a wrapper for zenoh-c and zenoh-pico and can work without any of these components. If zenoh-c is missing, the target zenohcxx::zenohc::lib is not defined too. See https://github.com/eclipse-zenoh/zenoh-cpp/blob/main/CMakeLists.txt#40, same logic is in the installed package.

There is issue about insufficient diagnostic when required library is not included: https://github.com/eclipse-zenoh/zenoh-cpp/issues/58

There are also few examples of simplest possible CMakeLists.txt's for zenoh-cpp: https://github.com/eclipse-zenoh/zenoh-cpp/tree/main/examples/simple

**Reply by lucasw** [reply]
Thanks, I got this working

https://github.com/lucasw/ros_one2z/blob/zenoh/ros1_example_pkg/CMakeLists.txt

after this install

https://github.com/lucasw/ros_one2z/tree/zenoh#install

Need to find_package on both zenohc and zenohcxx for the zenohcxx::zenohc::lib to work:
```
find_package(zenohc REQUIRED)
find_package(zenohcxx REQUIRED)
```
